src/write_open_telemetry.cc

```diff
@@ -0,0 +1,414 @@
+  metric_family_t *ret = NULL;
+  int status = c_avl_get(cb->staged_metric_families, fam->name, (void **)&ret);
+  if (status == 0) {
+    DEBUG("write_open_telemetry plugin: Found staged metric family \"%s\"",
+          ret->name);
+    return ret;
+  }
+
+  ret = metric_family_clone_shallow(fam);
+  c_avl_insert(cb->staged_metric_families, ret->name, ret);
+  DEBUG("write_open_telemetry plugin: Successfully staged metric family \"%s\"",
+        ret->name);
+  return ret;
+}
+
+static int ot_write(metric_family_t const *fam, user_data_t *user_data) {
+  if ((fam == NULL) || (user_data == NULL)) {
+    return EINVAL;
+  }
+
+  ot_callback_t *cb = (ot_callback_t *)user_data->data;
+  pthread_mutex_lock(&cb->mu);
+
+  if (ot_need_flush(cb, fam)) {
+    cdtime_t timeout = 0;
+    ot_flush_nolock(timeout, cb);
+  }
+
+  metric_family_t *stage = ot_staged_metric_family(cb, fam);
+  size_t offset = stage->metric.num;
+
+  int status = metric_family_append_list(stage, fam->metric);
+  if (status != 0) {
+    ERROR("write_open_telemetry plugin: metric_list_append_list failed: %d",
+          status);
+    pthread_mutex_unlock(&cb->mu);
+    return status;
+  }
+
+  for (size_t i = offset; i < stage->metric.num; i++) {
+    ot_mark_metric_staged(cb, &stage->metric.ptr[i]);
+  }
+
+  pthread_mutex_unlock(&cb->mu);
+  return 0;
+}
+
+static int ot_config_node(oconfig_item_t *ci) {
+  ot_callback_t *cb = (ot_callback_t *)calloc(1, sizeof(*cb));
+  if (cb == NULL) {
+    ERROR("write_open_telemetry plugin: calloc failed.");
+    return -1;
+  }
+
+  cb->reference_count = 1;
+  cf_util_get_string(ci, &cb->name);
+  cb->host = strdup(OT_DEFAULT_HOST);
+  cb->port = strdup(OT_DEFAULT_PORT);
+
+  cb->staged_metrics =
+      c_avl_create((int (*)(const void *, const void *))strcmp);
+  cb->staged_metric_families =
+      c_avl_create((int (*)(const void *, const void *))strcmp);
+
+  pthread_mutex_init(&cb->mu, /* attr = */ NULL);
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+
+    int status = 0;
+    if (strcasecmp("Host", child->key) == 0)
+      status = cf_util_get_string(child, &cb->host);
+    else if (strcasecmp("Port", child->key) == 0)
+      status = cf_util_get_service(child, &cb->port);
+    else {
+      ERROR("write_open_telemetry plugin: Invalid configuration "
+            "option: %s.",
+            child->key);
+      status = -1;
+    }
+
+    if (status != 0) {
+      ot_callback_decref(cb);
+      return status;
+    }
+  }
+
+  strbuf_t callback_name = STRBUF_CREATE;
+  strbuf_printf(&callback_name, "write_open_telemetry/%s", cb->name);
+
+  user_data_t user_data = {
+      .data = cb,
+      .free_func = ot_callback_decref,
+  };
+
+  cb->reference_count++;
+  plugin_register_write(callback_name.ptr, ot_write, &user_data);
+
+  cb->reference_count++;
+  plugin_register_flush(callback_name.ptr, ot_flush, &user_data);
```

Comment:
Btw. I'm not that familiar how plugin work.  Why these could not be registered in `module_register()`?

Reply:
This is the common pattern used when – depending on the configuration – multiple callbacks may be registered, e.g. to send metrics to multiple collectors. The information which callbacks to register is only available during the configuration phase, and therefore we can't register the callbacks in `module_register()`.